MapboxNavigationUI/RouteMapViewController.swift

```diff
@@ -240,7 +240,45 @@ extension RouteMapViewController: NavigationMapViewDelegate {
             }
         }
         
-        return CLLocation(coordinate: newCoordinate, altitude: location.altitude, horizontalAccuracy: location.horizontalAccuracy, verticalAccuracy: location.verticalAccuracy, course: location.course, speed: location.speed, timestamp: location.timestamp)
+        let defaultReturn = CLLocation(coordinate: newCoordinate, altitude: location.altitude, horizontalAccuracy: location.horizontalAccuracy, verticalAccuracy: location.verticalAccuracy, course: location.course, speed: location.speed, timestamp: location.timestamp)
+        
+        guard location.course != -1 else {
+            return defaultReturn
+        }
+        
+        let coords = routeController.routeProgress.currentLegProgress.nearbyCoordinates
+        assert(!coords.isEmpty)
+        
+        let closest = closestCoordinate(on: coords, to: location.coordinate)!
+        let slicedLine = polyline(along: coords, from: closest.coordinate, to: coords.last)
+        
+        
+        let userDistanceBuffer = location.speed * RouteControllerDeadReckoningTimeInterval
+        
+        // Get closest point infront of user
+        let infrontPointOne = coordinate(at: userDistanceBuffer, fromStartOf: slicedLine)!
+        let infrontClosest = closestCoordinate(on: coords, to: infrontPointOne)!
+        
+        // Get closest point behind in front of user
+        let infrontPointTwo = coordinate(at: userDistanceBuffer * 2, fromStartOf: slicedLine)!
+        let behindClosest = closestCoordinate(on: coords, to: infrontPointTwo)!
+        
+        // Get direction of these points
+        let infrontDirection = closest.coordinate.direction(to: infrontClosest.coordinate)
+        let behindDirection = closest.coordinate.direction(to: behindClosest.coordinate)
+        
+        let normalizedInfrontDirection = wrap(infrontDirection, min: 0, max: 360)
+        let normalizedBehindDirection = wrap(behindDirection, min: 0, max: 360)
+        
+        let diffBetweenAngleInFront = differenceBetweenAngles(location.course, normalizedInfrontDirection)
+        let diffBetweenAngleBehind = differenceBetweenAngles(location.course, normalizedBehindDirection)
+        
+        let averageDifference = (diffBetweenAngleInFront + diffBetweenAngleBehind) / 2
+        let averagedCourse = (normalizedBehindDirection + normalizedInfrontDirection) / 2
```

Comment:
If `location.course` is 0°, `normalizedInfrontDirection` is 1°, and `normalizedBehindDirection` is 359°, then `averagedDifference` is 0°, but `averagedCourse` is 179° – which is almost opposite of what you want.